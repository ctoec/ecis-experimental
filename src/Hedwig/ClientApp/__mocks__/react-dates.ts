const DayPickerSingleDateController = ({ ...props }) =>
	`<DayPickerSingleDateController" ${JSON.stringify(props)}/>`;

export { DayPickerSingleDateController };
