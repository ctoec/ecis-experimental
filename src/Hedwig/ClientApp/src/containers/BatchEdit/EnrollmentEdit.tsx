import React, { useContext } from 'react';
import UserContext from '../../contexts/User/UserContext';
import { ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGetRequest, Enrollment } from '../../generated';
import { getIdForUser } from '../../utils/models';
import useApi from '../../hooks/useApi';
import StepList, { StepProps } from '../../components/StepList/StepList';
import ChildInfo from './_sections/ChildInfo';
import { Section, SectionProps } from '../Enrollment/enrollmentTypes';
import { hasValidationErrors } from '../../utils/validations';

type EnrollmentEditProps = {
	enrollmentId: number;
	siteId: number;
}

export const EnrollmentEdit: React.FC<EnrollmentEditProps> = ({
	enrollmentId,
	siteId,
}) => {
	const { user } = useContext(UserContext);
	// get the enrollment detail
	const params: ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGetRequest = {
		id: enrollmentId,
		siteId,
		orgId: getIdForUser(user, 'org'),
	}

	const { error, data: enrollment, loading } = useApi<Enrollment>(
		(api) => api.apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet(params),
		{ skip: !user }
	)


	if (!enrollment) {
		return <>NOT FOUND</>;
	}

	const props: SectionProps = {
		enrollment,
		siteId,
		
	}
	const sections: Section[] = [];
	if(hasValidationErrors(enrollment.child, undefined, true)))
	const steps = sections.map((section) => {
		const editPath = section.key;
		return { ...section, editPath } as StepProps<SectionProps>
	})
	return (
		<StepList steps={steps} props={} />
	)
}
