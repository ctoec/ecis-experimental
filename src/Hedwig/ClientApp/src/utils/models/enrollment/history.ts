import { ProcessStepProps } from '../../../components/ProcessList/ProcessStep';
import { DeepNonUndefineable, DeepNonUndefineableArray } from '../../types';
import {
	Enrollment,
	Funding,
	FundingSource,
	FamilyDetermination,
	C4KCertificate,
} from '../../../generated';
import { prettyAge, prettyFundingTime, NO_FUNDING } from '..';
import dateFormatter from '../../dateFormatter';
import { dateSorter } from '../../dateSorter';
import moment from 'moment';
import { c4kCertificateSorter } from '../c4kCertificate';
import { fundingStartSorter, getFundingTime } from '../funding';

// sortable process step props include:
// - sortDate, the field to use for sorting
// - sortWeight, a number to use as a sorting tiebreaker to ensure enrollment starts come before funding/determination steps,
// 	and enrollment ends come after
type SortableProcessStepProps = ProcessStepProps & { sortDate: Date | null; sortWeight: number };

/**
 * Generates SortableProcessStepProps array for a given array of enrollments,
 * also creating step props for fundings and family determinations
 * @param enrollment
 * @returns sorted SortableProcessStepProps array
 */
export function getEnrollmentHistoryProps(
	currentEnrollment: DeepNonUndefineable<Enrollment>
): ProcessStepProps[] {
	// allEnrollments is sorted in descending order by Entry
	// (pastEnrollments are sorted thusly in API response)
	var allEnrollments = [
		currentEnrollment,
		...(currentEnrollment.pastEnrollments ? currentEnrollment.pastEnrollments : []),
	];

	var processStepProps: SortableProcessStepProps[] = [];

	allEnrollments.forEach((enrollment, idx) => {
		processStepProps = processStepProps.concat(
			enrollmentStepProps(enrollment, idx, allEnrollments.length)
		);

		// funding steps
		processStepProps = processStepProps.concat(
			fundingStepProps(enrollment.fundings, enrollment.exit)
		);

		// only add determination steps and C4K funding steps for the current enrollment (others will be dupes)
		if (idx === 0) {
			// determination steps
			processStepProps = processStepProps.concat(
				determinationStepProps(enrollment.child.family.determinations)
			);

			// C4K certification steps
			processStepProps = processStepProps.concat(
				c4kCertificateStepProps(enrollment.child.c4KCertificates)
			);
		}
	});

	// return all steps sorted by `sortDate`, and sortWeight
	return processStepProps.sort(enrollmentHistoryPropsSorter);
}

export function enrollmentStepProps(
	_enrollment: DeepNonUndefineable<Enrollment>,
	enrollmentIndex: number,
	allEnrollmentsCount: number
) {
	var processStepProps: SortableProcessStepProps[] = [];

	// enrollment start step
	if (_enrollment.entry) {
		processStepProps.push(
			getSortableStep({
				heading:
					enrollmentIndex === allEnrollmentsCount - 1
						? // heading text for first enrollment (last in ordered array)
						  `Enrolled in ${_enrollment.site.name}`
						: `Changed enrollment to ${prettyAge(_enrollment.ageGroup)} in ${
								_enrollment.site.name
						  }`,
				body: `on ${dateFormatter(_enrollment.entry, false)}`,
				stepDate: _enrollment.entry,
				stepWeight: -1,
			})
		);
	}

	// enrollment stop step (only care about withdrawal for the last enrollment)
	if (enrollmentIndex === 0 && _enrollment.exit) {
		processStepProps.push(
			getSortableStep({
				heading: `Withdrawn from ${_enrollment.site.name}`,
				body: `on ${dateFormatter(_enrollment.exit, false)}`,
				stepDate: _enrollment.exit,
				stepWeight: 1,
			})
		);
	}

	return processStepProps;
}

/**
 * Generates SortableProcessStepProps array for a given array of fundings
 * @param fundings
 * @param enrollmentExit the value of Exit on the owning enrollment,
 * used to determine if switch to private pay step should be created
 */
export function fundingStepProps(
	fundings: DeepNonUndefineable<Funding[]> | null,
	enrollmentExit: Date | null
) {
	var processStepProps: SortableProcessStepProps[] = [];

	if (fundings) {
		fundings.sort(fundingStartSorter).forEach(funding => {
			// CDC Funding start if:
			// - source is CDC
			// - funding has first reporting period
			if (funding.source === FundingSource.CDC && funding.firstReportingPeriod) {
				var heading = `Switched funding to ${funding.source}`;
				var fundingTime = getFundingTime(funding);
				if (fundingTime) heading += ` - ${prettyFundingTime(fundingTime)}`;

				processStepProps.push(
					getSortableStep({
						heading,
						body: `on ${dateFormatter(funding.firstReportingPeriod.periodStart, false)}`,
						stepDate: funding.firstReportingPeriod.periodStart,
					})
				);
			}

			// Switch to private pay if:
			// - source is CDC
			// - funding has last reporting period
			// - enrollment is still ongoing (exit == null)
			// 	 OR last reporting period ends before enrollment exit
			// TODO: Handle other funding sources
			if (
				funding.source === FundingSource.CDC &&
				funding.lastReportingPeriod &&
				(!enrollmentExit || funding.lastReportingPeriod.periodEnd < enrollmentExit)
			) {
				var dayAfterLastReportingPeriodEnd = moment(funding.lastReportingPeriod.periodEnd)
					.add(1, 'days')
					.toDate();
				processStepProps.push(
					getSortableStep({
						heading: `Switched funding to ${NO_FUNDING}`,
						body: `on ${dateFormatter(dayAfterLastReportingPeriodEnd, false)}`,
						stepDate: dayAfterLastReportingPeriodEnd
					})
				);
			}
		});
	}

	return processStepProps;
}

/**
 * Generates SortableProcessStepProps array for a given array of family determinations
 * @param determinations
 */
export function determinationStepProps(
	determinations: DeepNonUndefineable<FamilyDetermination[]> | null
) {
	var processStepProps: SortableProcessStepProps[] = [];
	if (determinations) {
		determinations.forEach(determination => {
			if (!determination.notDisclosed && determination.determinationDate) {
				processStepProps.push(
					getSortableStep({
						heading: 'Income redetermined',
						body: `on ${dateFormatter(determination.determinationDate, false)}`,
						stepDate: determination.determinationDate,
					})
				);
			}
		});
	}

	return processStepProps;
}

/**
 * Generates SortableProcessStepProps array for a given array of C4KCertificates
 * @param c4KCertificates
 */
export function c4kCertificateStepProps(
	c4KCertificates: DeepNonUndefineable<C4KCertificate[]> | null
) {
	var processStepProps: SortableProcessStepProps[] = [];

	if (c4KCertificates) {
		c4KCertificates.sort(c4kCertificateSorter).forEach((c4kCertificate, idx) => {
			if (c4kCertificate.startDate) {
				processStepProps.push(
					getSortableStep({
						heading:
							idx === 0 ? 'Care 4 Kids certificate added' : 'Care 4 Kids certificate renewed',
						body: `on ${dateFormatter(c4kCertificate.startDate, false)}`,
						stepDate: c4kCertificate.startDate,
					})
				);
			}
		});
	}

	return processStepProps;
}

/**
 * Sorts SortableProcessStepProp objects inversely by sortDate, using sortWeight as a tie breaker
 * @param a
 * @param b
 */
export function enrollmentHistoryPropsSorter(
	a: SortableProcessStepProps,
	b: SortableProcessStepProps
) {
	var dateSorterRes = dateSorter(a.sortDate, b.sortDate, true);

	if (dateSorterRes !== 0) return dateSorterRes;

	if (a.sortWeight < b.sortWeight) return 1;
	if (a.sortWeight > b.sortWeight) return -1;
	return 0;
}

/**
 * Returns a SortableProcessStepProps object composed from the given parameters,
 * using the stepDate as sortDate, and to get correct value for isNew
 * @param heading
 * @param body
 * @param stepDate
 * @param stepWeight
 */
export function getSortableStep(params: {
	heading: string;
	body: string;
	stepDate: Date;
	stepWeight?: number;
}): SortableProcessStepProps {
	return {
		heading: params.heading,
		body: params.body,
		sortDate: params.stepDate,
		sortWeight: params.stepWeight ? params.stepWeight : 0,
		isNew: stepIsNew(params.stepDate),
	};
}

/**
 * Returns true is date is less than 1 week old, else false
 * @param stepDate
 */
export function stepIsNew(stepDate: Date) {
	return moment().add(-1, 'week') < moment(stepDate);
}
