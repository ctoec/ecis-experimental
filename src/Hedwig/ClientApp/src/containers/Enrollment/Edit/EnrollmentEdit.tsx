import React, { useContext } from 'react';
import { Section } from '../enrollmentTypes';
import { History } from 'history';
import ChildInfo from '../_sections/ChildInfo';
import FamilyInfo from '../_sections/FamilyInfo';
import FamilyIncome from '../_sections/FamilyIncome';
import EnrollmentFunding from '../_sections/EnrollmentFunding';
import PageNotFound from '../../PageNotFound/PageNotFound';
import UserContext from '../../../contexts/User/UserContext';
import {
	Enrollment,
	ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGetRequest,
} from '../../../generated';
import getIdForUser from '../../../utils/getIdForUser';
import useApi from '../../../hooks/useApi';
import CommonContainer from '../../CommonContainer';
import { hasValidationErrors } from '../../../utils/validations';
import AlertContext from '../../../contexts/Alert/AlertContext';
import nameFormatter from '../../../utils/nameFormatter';

type EnrollmentEditParams = {
	history: History;
	match: {
		params: {
			enrollmentId: number;
			sectionId: string;
		};
	};
};

const sections: { [key: string]: Section } = {
	'child-information': ChildInfo,
	'family-information': FamilyInfo,
	'family-income': FamilyIncome,
	'enrollment-funding': EnrollmentFunding,
};

/**
 * React component for editing an enrollment. Hands off to a section
 * form component.
 *
 * @param props Props with location.
 */
export default function EnrollmentEdit({
	history,
	match: {
		params: { enrollmentId, sectionId },
	},
}: EnrollmentEditParams) {
	const section = sections[sectionId];
	const { user } = useContext(UserContext);
	const { setAlerts } = useContext(AlertContext);

	// Get enrollment by id
	const params: ApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGetRequest = {
		id: enrollmentId ? enrollmentId : 0,
		orgId: getIdForUser(user, 'org'),
		siteId: getIdForUser(user, 'site'),
		include: ['child', 'family', 'determinations', 'fundings'],
	};
	const [loading, error, enrollment, mutate] = useApi(
		api => api.apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet(params),
		[user]
	);

	if (!section) {
		return <PageNotFound />;
	}

	if (loading || error || !enrollment) {
		return <div className="EnrollmentEdit"></div>;
	}

	/**
	 * Accepts an enrollment and navigates back to the enrollment
	 * summary page.
	 *
	 * @param enrollment Enrollment that was just saved.
	 */
	const afterSave = (enrollment: Enrollment) => {
		const childName = nameFormatter(enrollment.child);
		const inSiteName = enrollment.site ? ` in ${enrollment.site.name}` : '';
		let successAlertText = `${childName}'s enrollment${inSiteName} has been updated.`;
		const informationIsMissing = hasValidationErrors(enrollment);
		if (informationIsMissing) {
			successAlertText +=
				' However, there is missing information you are required to enter before you can submit your monthly CDC report.';
		}
		setAlerts([
			{
				type: 'success',
				heading: 'Enrolled',
				text: successAlertText,
			},
		]);
		history.push(`/roster/enrollments/${enrollment.id}/`);
	};

	return (
		<CommonContainer
			directionalLinkProps={{
				direction: 'left',
				to: `/roster/enrollments/${enrollment.id}/`,
				text: `Back to enrollment details`,
			}}
		>
			<section className="grid-container">
				<h1>Edit {section.name.toLowerCase()}</h1>
				<section.Form enrollment={enrollment} mutate={mutate} successCallback={afterSave} />
			</section>
		</CommonContainer>
	);
}
