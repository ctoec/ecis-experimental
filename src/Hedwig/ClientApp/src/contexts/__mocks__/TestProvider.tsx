import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, History } from 'history';
import AuthenticationContext from '../Authentication/AuthenticationContext';
import UserContext from '../User/UserContext';
import ReportingPeriodContext from '../ReportingPeriod/ReportingPeriodContext';
import { User, ReportingPeriod } from '../../generated';
import { cdcReportingPeriods as _cdcReportingPeriods, user as _user } from '../../tests/data';

type TestProviderProps = {
	user?: User;
	cdcReportingPeriods?: ReportingPeriod[];
	history?: History<any>;
};

const TestProvider: React.FC<TestProviderProps> = ({
	children,
	user = _user,
	cdcReportingPeriods = _cdcReportingPeriods,
	history = createMemoryHistory(),
}) => {
	return (
		<AuthenticationContext.Provider
			value={{ accessToken: '', withFreshToken: jest.fn(), loading: false }}
		>
			<UserContext.Provider value={{ user, loading: false }}>
				<ReportingPeriodContext.Provider value={{ cdcReportingPeriods }}>
					<Router history={history}>{children}</Router>
				</ReportingPeriodContext.Provider>
			</UserContext.Provider>
		</AuthenticationContext.Provider>
	);
};

export default TestProvider;
