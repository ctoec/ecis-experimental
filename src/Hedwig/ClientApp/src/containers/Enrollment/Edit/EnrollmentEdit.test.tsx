// Variables used in jest mockes -- must start with `mock`
import { mockAllFakeEnrollments, mockSite } from '../../../tests/data';
import mockUseApi, {
	mockApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet,
	mockApiOrganizationsOrgIdSitesIdGet,
} from '../../../hooks/__mocks__/useApi';

// Jest mocks must occur before later imports
jest.mock('../../../hooks/useApi', () =>
	mockUseApi({
		apiOrganizationsOrgIdSitesIdGet: mockApiOrganizationsOrgIdSitesIdGet(mockSite),
		apiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet: mockApiOrganizationsOrgIdSitesSiteIdEnrollmentsIdGet(
			mockAllFakeEnrollments
		),
	})
);

import React from 'react';
import mockdate from 'mockdate';
import { createBrowserHistory } from 'history';
import { render, getAllByRole } from '@testing-library/react';
import 'react-dates/initialize';
import TestProvider from '../../../contexts/__mocks__/TestProvider';
import EnrollmentEdit from './EnrollmentEdit';
import { accessibilityTestHelper } from '../../../tests/helpers';
import {
	mockCompleteEnrollment,
	cdcReportingPeriods,
	mockEnrollmentMissingAddress,
} from '../../../tests/data';

const fakeDate = '2019-03-02';

beforeAll(() => {
	mockdate.set(fakeDate);
});

afterAll(() => {
	mockdate.reset();
	jest.resetModules();
});

const history = createBrowserHistory();

describe('EnrollmentEdit', () => {
	describe('family info', () => {
		it('shows a fieldset warning if there is no address', () => {
			const { getByRole } = render(
				<TestProvider>
					<EnrollmentEdit
						history={history}
						match={{
							params: {
								siteId: mockEnrollmentMissingAddress.siteId,
								enrollmentId: mockEnrollmentMissingAddress.id,
								sectionId: 'family-information',
							},
						}}
					/>
				</TestProvider>
			);

			const addressErr = getByRole('status');

			expect(addressErr.classList.contains('usa-warning-message'));
		});
	});

	describe('family income', () => {
		it('shows an info alert if family income is not disclosed', () => {
			const { getByText } = render(
				<TestProvider>
					<EnrollmentEdit
						history={history}
						match={{
							params: {
								siteId: mockCompleteEnrollment.siteId,
								enrollmentId: mockCompleteEnrollment.id,
								sectionId: 'family-income',
							},
						}}
					/>
				</TestProvider>
			);

			const infoAlert = getByText(
				'Income information is required to enroll a child in a CDC funded space. You will not be able to assign this child to a funding space without this information.'
			);

			expect(infoAlert).toBeInTheDocument();
		});
	});

	describe('enrollment and funding', () => {
		it('shows the appropriate number of reporting periods for enrollment funding', () => {
			const { getByLabelText } = render(
				<TestProvider>
					<EnrollmentEdit
						history={history}
						match={{
							params: {
								siteId: mockCompleteEnrollment.siteId,
								enrollmentId: mockCompleteEnrollment.id,
								sectionId: 'enrollment-funding',
							},
						}}
					/>
				</TestProvider>
			);
			const reportingPeriodSelect = getByLabelText('First reporting period');
			const reportingPeriodOptions = getAllByRole(reportingPeriodSelect, 'option');
			expect(reportingPeriodOptions.length).toBe(cdcReportingPeriods.length + 1);
		});
	});

	accessibilityTestHelper(
		<TestProvider>
			<EnrollmentEdit
				history={history}
				match={{
					params: {
						siteId: mockCompleteEnrollment.siteId,
						enrollmentId: mockCompleteEnrollment.id,
						sectionId: 'enrollment-funding',
					},
				}}
			/>
		</TestProvider>
	);
});
