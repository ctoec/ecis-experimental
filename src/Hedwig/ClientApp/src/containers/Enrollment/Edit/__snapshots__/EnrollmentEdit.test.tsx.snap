// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnrollmentEdit enrollment and funding matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid-container"
  >
    <a
      class="oec-directional-link"
      href="/roster/sites/1/enrollments/1/"
    >
      <svg
        class="oec-directional-link__arrow oec-directional-link__arrow--direction-left oec-directional-link__arrow--side-left"
      >
        arrowRight.svg
      </svg>
      Back to enrollment details
    </a>
  </div>
  <div
    class="grid-container"
  >
    <h1>
      Edit enrollment and funding
    </h1>
    <p
      class="usa-intro"
    >
      Lily Luna Potter
    </p>
    <div
      class="usa-alert usa-alert--error grid-row margin-y-4"
      role="alert"
    >
      <div
        class="display-flex flex-fill"
      >
        <div
          class="usa-alert__body grid-col flex-fill"
        >
          <h2
            class="usa-alert__heading"
          >
            Something went wrong
          </h2>
          <p
            class="usa-alert__text"
          >
            <span>
              We were unable to save your changes. Please try again. If you continue to experience problems, contact 
              <a
                href="mailto:oec-data-pilot@skylight.digital&subject=ECE%20reporter%20bug"
                rel="noopener noreferrer"
                target="_blank"
              >
                oec-data-pilot@skylight.digital
              </a>
               for help.
            </span>
          </p>
        </div>
        <div
          class="usa-alert__body flex-auto display-flex flex-align-center"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
